src/BenchmarkDotNet/Engines/StoppingCriteria/StoppingCriteriaBase.cs

```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using BenchmarkDotNet.Reports;
+using JetBrains.Annotations;
+
+namespace BenchmarkDotNet.Engines
+{
+    public abstract class StoppingCriteriaBase : IStoppingCriteria
+    {
+        private readonly Lazy<string> lazyTitle;
+        private readonly Lazy<int> lazyMaxIterationCount;
+        private readonly Lazy<string[]> lazyWarnings;
+
+        public string Title => lazyTitle.Value;
+        public int MaxIterationCount => lazyMaxIterationCount.Value;
+        public string[] Warnings => lazyWarnings.Value;
```

Comment:
nit: I think that if we have methods like `GetXYZ` we don't need `XYZ` property.

Reply:
In this case, all of the `GetXYZ` methods are protected while the `XYZ` properties are public API. We need them to fill the private lazy fields which are used in the property getters. This approach helps to control that all of the properties are actually lazy.